Tests using `invoke_and_assert` can sometimes attempt to forward signals from a thread that's not `main`

### First check

- [X] I am a contributor to the Prefect codebase

### Description

This was the source of several test flakes that @abrookins identified here: https://github.com/PrefectHQ/prefect/pull/10889/files#diff-00c0a3f9f0ef58527f83db91f74ea7356bb8d9134b4db3e490e4e35d39c49445R12-R16

### Impact

_No response_

### Additional context

_No response_

Here's a recent example of this failure -- it can occur in many places, not just our CLI commands.

My problem with this is that the issue is not only that tests are flaky -- though they are. The problem is that this suggests any code wrapped in `sync_compatible` is **not safe to use with signal handlers**.

```
=================================== FAILURES ===================================
________________________ test_start_no_options[server] _________________________
[gw7] linux -- Python 3.11.5 /opt/hostedtoolcache/Python/3.11.5/x64/bin/python

mock_run_process = <AsyncMock id='140453732195280'>, command_group = 'server'

    @pytest.mark.parametrize("command_group", ["server"])
    def test_start_no_options(mock_run_process: AsyncMock, command_group: str):
>       invoke_and_assert(
            [command_group, "start"],
            expected_output_contains=[
                "prefect config set PREFECT_API_URL=http://127.0.0.1:4200/api",
                "View the API reference documentation at http://127.0.0.1:4200/docs",
            ],
        )

tests/cli/test_server.py:26: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/prefect/testing/cli.py:99: in invoke_and_assert
    result = runner.invoke(app, command, catch_exceptions=False, input=user_input)
/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/typer/testing.py:21: in invoke
    return super().invoke(
/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/click/testing.py:408: in invoke
    return_value = cli.main(args=args or (), prog_name=prog_name, **extra)
/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/typer/core.py:778: in main
    return _main(
/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/typer/core.py:216: in _main
    rv = self.invoke(ctx)
/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/click/core.py:1688: in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/click/core.py:1688: in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/click/core.py:1434: in invoke
    return ctx.invoke(self.callback, **ctx.params)
/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/click/core.py:783: in invoke
    return __callback(*args, **kwargs)
/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/site-packages/typer/main.py:683: in wrapper
    return callback(**use_params)  # type: ignore
src/prefect/cli/_utilities.py:41: in wrapper
    return fn(*args, **kwargs)
src/prefect/utilities/asyncutils.py:255: in coroutine_wrapper
    return call()
src/prefect/_internal/concurrency/calls.py:382: in __call__
    return self.result()
src/prefect/_internal/concurrency/calls.py:282: in result
    return self.future.result(timeout=timeout)
src/prefect/_internal/concurrency/calls.py:168: in result
    return self.__get_result()
/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/concurrent/futures/_base.py:401: in __get_result
    raise self._exception
src/prefect/_internal/concurrency/calls.py:345: in _run_async
    result = await coro
src/prefect/cli/server.py:151: in start
    setup_signal_handlers_server(
src/prefect/utilities/processutils.py:364: in setup_signal_handlers_server
    setup_handler(signal.SIGINT, signal.SIGTERM, signal.SIGKILL)
src/prefect/utilities/processutils.py:349: in forward_signal_handler
    signal.signal(signum, handler)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

signalnum = <Signals.SIGINT: 2>
handler = <function forward_signal_handler.<locals>.handler at 0x7fbdd0eda480>

    @_wraps(_signal.signal)
    def signal(signalnum, handler):
>       handler = _signal.signal(_enum_to_int(signalnum), _enum_to_int(handler))
E       ValueError: signal only works in main thread of the main interpreter

/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/signal.py:56: ValueError
----------------------------- Captured stderr call -----------------------------
19:25:51.375 | DEBUG   | Dummy-1      | prefect._internal.concurrency - <function start at 0x7fbe18a21620> --> run async in new loop
19:25:51.375 | DEBUG   | Dummy-1      | prefect._internal.concurrency - Running call start(host='127.0.0.1', port=4200, keep_alive_timeout=5, log_level='WARNING', scheduler=False, analytics=F...) in thread 'Dummy-1'
19:25:51.375 | DEBUG   | Dummy-1      | prefect._internal.concurrency - <WatcherThreadCancelScope, name='start' RUNNING, runtime=0.00> entered
19:25:51.375 | DEBUG   | Dummy-1      | prefect._internal.concurrency - <WatcherThreadCancelScope, name='start' COMPLETED, runtime=0.00> exited
19:25:51.375 | DEBUG   | Dummy-1      | prefect._internal.concurrency - Executing coroutine for call start(host='127.0.0.1', port=4200, keep_alive_timeout=5, log_level='WARNING', scheduler=False, analytics=F...) in new loop
19:25:51.376 | DEBUG   | Dummy-1      | prefect._internal.concurrency - <AsyncCancelScope, name='start' RUNNING, runtime=0.00> entered
19:25:51.379 | DEBUG   | Dummy-1      | prefect._internal.concurrency - <AsyncCancelScope, name='start' COMPLETED, runtime=0.00> exited
19:25:51.379 | DEBUG   | Dummy-1      | prefect._internal.concurrency - Encountered exception in async call start(host='127.0.0.1', port=4200, keep_alive_timeout=5, log_level='WARNING', scheduler=False, analytics=F...)
Traceback (most recent call last):
  File "/home/runner/work/prefect/prefect/src/prefect/_internal/concurrency/calls.py", line 345, in _run_async
    result = await coro
             ^^^^^^^^^^
  File "/home/runner/work/prefect/prefect/src/prefect/cli/server.py", line 151, in start
    setup_signal_handlers_server(
  File "/home/runner/work/prefect/prefect/src/prefect/utilities/processutils.py", line 364, in setup_signal_handlers_server
    setup_handler(signal.SIGINT, signal.SIGTERM, signal.SIGKILL)
  File "/home/runner/work/prefect/prefect/src/prefect/utilities/processutils.py", line 349, in forward_signal_handler
    signal.signal(signum, handler)
  File "/opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/signal.py", line 56, in signal
    handler = _signal.signal(_enum_to_int(signalnum), _enum_to_int(handler))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: signal only works in main thread of the main interpreter
```

--- a/tests/public/flows/test_flow_crashes.py
+++ b/tests/public/flows/test_flow_crashes.py
@@ -9,6 +9,7 @@ interrupts.
 import asyncio
 import os
 import signal
+import threading
 from unittest.mock import ANY, MagicMock
 
 import anyio
@@ -211,6 +212,8 @@ async def test_interrupt_in_child_crashes_parent_and_child_flow(
 
 @pytest.fixture
 def mock_sigterm_handler():
+    if threading.current_thread() != threading.main_thread():
+        pytest.skip("Can't test signal handlers from a thread")
     mock = MagicMock()
 
     def handler(*args, **kwargs):

--- a/tests/test_flows.py
+++ b/tests/test_flows.py
@@ -5,6 +5,7 @@ import inspect
 import os
 import signal
 import sys
+import threading
 import time
 from functools import partial
 from itertools import combinations
@@ -82,6 +83,8 @@ def test_flow():
 
 @pytest.fixture
 def mock_sigterm_handler():
+    if threading.current_thread() != threading.main_thread():
+        pytest.skip("Can't test signal handlers from a thread")
     mock = MagicMock()
 
     def handler(*args, **kwargs):

